File: internal/docgen/docgen.go
package docgen

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"sync"
	"text/template"

	"github.com/delightfulhammers/telesis/templates"
)

var (
	slugRe          = regexp.MustCompile(`^[a-z0-9]+(?:-[a-z0-9]+)*$`)
	parsedTemplates *template.Template
	parseOnce       sync.Once
	parseError      error
)

func loadTemplates() (*template.Template, error) {
	parseOnce.Do(func() {
		parsedTemplates, parseError = template.ParseFS(templates.FS, "*.tmpl")
	})
	return parsedTemplates, parseError
}

// Config describes a numbered document type (ADR, TDD, etc.).
type Config struct {
	Prefix   string // e.g., "ADR" or "TDD"
	Subdir   string // e.g., "adr" or "tdd"
	Template string // e.g., "adr.md.tmpl"
}

// TemplateData is passed to the document template.
type TemplateData struct {
	Number int
	Slug   string
	Padded string
}

// Create creates a new numbered document with collision-safe file creation.
// It returns the path to the created file.
func Create(rootDir string, cfg Config, slug string) (string, error) {
	if err := ValidateSlug(slug); err != nil {
		return "", err
	}

	docDir := filepath.Join(rootDir, "docs", cfg.Subdir)
	numberRe := regexp.MustCompile(`^` + cfg.Prefix + `-(\d+)`)

	const maxRetries = 5
	for attempt := range maxRetries {
		num, err := nextNumber(docDir, numberRe)
		if err != nil {
			return "", fmt.Errorf("determining next %s number: %w", cfg.Prefix, err)
		}

		content, err := render(cfg.Template, num, slug)
		if err != nil {
			return "", err
		}

		filename := fmt.Sprintf("%s-%03d-%s.md", cfg.Prefix, num, slug)
		dest := filepath.Join(docDir, filename)

		err = writeExclusive(dest, content)
		if err == nil {
			return dest, nil
		}
		if !errors.Is(err, os.ErrExist) {
			return "", fmt.Errorf("writing %s: %w", filename, err)
		}
		// File already exists (concurrent creation) — retry with new number
		_ = attempt
	}

	return "", fmt.Errorf("could not create %s after %d attempts (concurrent collision)", cfg.Prefix, maxRetries)
}

// NextNumber scans a document directory and returns the next sequential number.
func NextNumber(docDir string, prefix string) (int, error) {
	numberRe := regexp.MustCompile(`^` + prefix + `-(\d+)`)
	return nextNumber(docDir, numberRe)
}

func nextNumber(docDir string, numberRe *regexp.Regexp) (int, error) {
	entries, err := os.ReadDir(docDir)
	if err != nil {
		return 0, fmt.Errorf("reading directory: %w", err)
	}

	highest := 0
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		if m := numberRe.FindStringSubmatch(entry.Name()); len(m) > 1 {
			n, _ := strconv.Atoi(m[1])
			if n > highest {
				highest = n
			}
		}
	}

	return highest + 1, nil
}

// ValidateSlug checks that a slug is valid kebab-case.
func ValidateSlug(slug string) error {
	if slug == "" {
		return fmt.Errorf("slug is required")
	}
	if !slugRe.MatchString(slug) {
		return fmt.Errorf("slug must be lowercase alphanumeric with hyphens (e.g., 'use-nats-for-events')")
	}
	return nil
}

func render(templateName string, num int, slug string) ([]byte, error) {
	tmpls, err := loadTemplates()
	if err != nil {
		return nil, fmt.Errorf("loading templates: %w", err)
	}

	data := TemplateData{
		Number: num,
		Slug:   slug,
		Padded: fmt.Sprintf("%03d", num),
	}

	var buf bytes.Buffer
	if err := tmpls.ExecuteTemplate(&buf, templateName, data); err != nil {
		return nil, fmt.Errorf("rendering template %s: %w", templateName, err)
	}

	return buf.Bytes(), nil
}

func writeExclusive(dest string, content []byte) error {
	f, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o666)
	if err != nil {
		return err
	}
	defer f.Close()

	if _, err := f.Write(content); err != nil {
		os.Remove(dest)
		return err
	}

	return nil
}

File: src/docgen/docgen.ts
import {
  readdirSync,
  writeFileSync,
  openSync,
  closeSync,
  unlinkSync,
  constants,
  type Dirent,
} from "node:fs";
import { join } from "node:path";
import { renderTemplate } from "../templates/index.js";
import type { TemplateName } from "../templates/index.js";

const SLUG_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

export interface DocConfig {
  readonly prefix: string; // e.g., "ADR" or "TDD"
  readonly subdir: string; // e.g., "adr" or "tdd"
  readonly template: TemplateName; // e.g., "adr.md.tmpl"
}

export const validateSlug = (slug: string): void => {
  if (!slug) {
    throw new Error("slug is required");
  }
  if (!SLUG_RE.test(slug)) {
    throw new Error(
      "slug must be lowercase alphanumeric with hyphens (e.g., 'use-nats-for-events')",
    );
  }
};

export const nextNumber = (docDir: string, prefix: string): number => {
  const numberRe = new RegExp(`^${escapeRegExp(prefix)}-(\\d+)`);

  let entries: Dirent[];
  try {
    entries = readdirSync(docDir, { withFileTypes: true });
  } catch {
    throw new Error(`reading directory: ${docDir}`);
  }

  let highest = 0;
  for (const entry of entries) {
    if (entry.isDirectory()) continue;
    const name = entry.name;
    const m = numberRe.exec(name);
    if (m?.[1]) {
      const n = parseInt(m[1], 10);
      if (n > highest) highest = n;
    }
  }

  return highest + 1;
};

const escapeRegExp = (s: string): string =>
  s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

const renderDoc = (
  templateName: TemplateName,
  num: number,
  slug: string,
): string =>
  renderTemplate(templateName, {
    Number: num,
    Slug: slug,
    Padded: String(num).padStart(3, "0"),
  });

const writeExclusive = (dest: string, content: string): void => {
  // O_EXCL ensures the file doesn't already exist
  let fd: number;
  try {
    fd = openSync(
      dest,
      constants.O_WRONLY | constants.O_CREAT | constants.O_EXCL,
      0o666,
    );
  } catch (err) {
    const e = err as NodeJS.ErrnoException;
    if (e.code === "EEXIST") {
      throw Object.assign(new Error(`file already exists: ${dest}`), {
        code: "EEXIST",
      });
    }
    throw err;
  }

  try {
    writeFileSync(fd, content);
  } catch (writeErr) {
    closeSync(fd);
    try {
      unlinkSync(dest);
    } catch {
      // cleanup best-effort
    }
    throw writeErr;
  }

  closeSync(fd);
};

const hasSequenceConflict = (
  docDir: string,
  prefix: string,
  num: number,
  ownFilename: string,
): boolean => {
  const padded = String(num).padStart(3, "0");
  const seqPrefix = `${prefix}-${padded}-`;
  const entries = readdirSync(docDir, { withFileTypes: true });

  return entries.some(
    (entry) =>
      !entry.isDirectory() &&
      entry.name !== ownFilename &&
      entry.name.startsWith(seqPrefix),
  );
};

export const create = (
  rootDir: string,
  cfg: DocConfig,
  slug: string,
): string => {
  validateSlug(slug);

  const docDir = join(rootDir, "docs", cfg.subdir);
  const MAX_RETRIES = 5;

  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    const num = nextNumber(docDir, cfg.prefix);
    const content = renderDoc(cfg.template, num, slug);
    const filename = `${cfg.prefix}-${String(num).padStart(3, "0")}-${slug}.md`;
    const dest = join(docDir, filename);

    try {
      writeExclusive(dest, content);
    } catch (err) {
      const e = err as NodeJS.ErrnoException;
      if (e.code === "EEXIST") {
        continue;
      }
      throw new Error(`writing ${filename}: ${(err as Error).message}`);
    }

    // Post-write check: verify no other file claimed the same sequence number
    if (hasSequenceConflict(docDir, cfg.prefix, num, filename)) {
      try {
        unlinkSync(dest);
      } catch {
        // cleanup best-effort
      }
      continue;
    }

    return dest;
  }

  throw new Error(
    `could not create ${cfg.prefix} after ${MAX_RETRIES} attempts (concurrent collision)`,
  );
};
